### PR TITLE
Transformations fix

### PR DESCRIPTION
**Pull request by costashatz**
Proper transformations for position/angle computations + fix for https://github.com/resibots/hexapod_common/pull/5 .

### CODE REVIEW

File: hexapod_dart/include/hexapod_dart/descriptors.hpp [hunk 1]
```diff
@@ -75,9 +75,19 @@ namespace hexapod_dart {
...
-                _pos_traj.push_back(rob->pos() - init_pos);
+                Eigen::Vector6d pose = rob->pose();
+                Eigen::Matrix3d rr = dart::math::expMapRot({pose[0], pose[1], pose[2]});
+                Eigen::Matrix3d ro = dart::math::expMapRot({init_trans[0], init_trans[1], init_trans[2]});
+                Eigen::MatrixXd init(4, 4);
+                init << ro(0, 0), ro(0, 1), ro(0, 2), init_trans[3], ro(1, 0), ro(1, 1), ro(1, 2), init_trans[4], ro(2, 0), ro(2, 1), ro(2, 2), init_trans[5], 0, 0, 0, 1;
+                Eigen::MatrixXd pp(4, 4);
+                pp << rr(0, 0), rr(0, 1), rr(0, 2), pose[3], rr(1, 0), rr(1, 1), rr(1, 2), pose[4], rr(2, 0), rr(2, 1), rr(2, 2), pose[5], 0, 0, 0, 1;
+                Eigen::Vector4d p = {init_trans[3], init_trans[4], init_trans[5], 1.0};
+                p = init.inverse() * pp * p;
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/descriptors.hpp]: If it's relevant, we can consider avoiding the inverse transform.

File: hexapod_dart/include/hexapod_dart/descriptors.hpp [hunk 2]
```diff
@@ -94,11 +104,12 @@ namespace hexapod_dart {
             RotationTraj() {}
 
             template <typename Simu, typename robot>
-            void operator()(Simu& simu, std::shared_ptr<robot> rob, const Eigen::Vector3d& init_pos, const Eigen::Vector3d& init_rot)
+            void operator()(Simu& simu, std::shared_ptr<robot> rob, const Eigen::Vector6d& init_trans)
             {
                 // roll-pitch-yaw
-                auto rot_mat = dart::math::expMapRot(rob->rot() - init_rot);
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/descriptors.hpp]: How does the `rob->rot() - init_rot` work ? What does it give you ?

File: hexapod_dart/include/hexapod_dart/descriptors.hpp [hunk 3]
```diff
@@ -94,11 +104,12 @@ namespace hexapod_dart {
             RotationTraj() {}
 
             template <typename Simu, typename robot>
-            void operator()(Simu& simu, std::shared_ptr<robot> rob, const Eigen::Vector3d& init_pos, const Eigen::Vector3d& init_rot)
+            void operator()(Simu& simu, std::shared_ptr<robot> rob, const Eigen::Vector6d& init_trans)
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/descriptors.hpp:107]: `init_trans` reads _init translation_ in my head but I assume that you mean _init transformation_

File: hexapod_dart/include/hexapod_dart/hexapod.hpp [hunk 4]
```diff
@@ -72,7 +72,7 @@ namespace hexapod_dart {
         {
             auto pos_and_rot = _skeleton->getPositions();
             Eigen::Vector6d tmp;
-            tmp << pos_and_rot(3), pos_and_rot(4), pos_and_rot(5), pos_and_rot(0), pos_and_rot(1), pos_and_rot(2);
+            tmp << pos_and_rot(0), pos_and_rot(1), pos_and_rot(2), pos_and_rot(3), pos_and_rot(4), pos_and_rot(5);
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/hexapod.hpp:78]: What is this line doing ?

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp [hunk 5]
```diff
@@ -161,16 +159,26 @@ namespace hexapod_dart {
...
+            // Position computation
+            Eigen::Vector6d pose = rob->pose();
+            Eigen::Matrix3d rr = dart::math::expMapRot({pose[0], pose[1], pose[2]});
+            Eigen::Matrix3d ro = dart::math::expMapRot({init_trans[0], init_trans[1], init_trans[2]});
+            Eigen::MatrixXd init(4, 4);
+            init << ro(0, 0), ro(0, 1), ro(0, 2), init_trans[3], ro(1, 0), ro(1, 1), ro(1, 2), init_trans[4], ro(2, 0), ro(2, 1), ro(2, 2), init_trans[5], 0, 0, 0, 1;
+            Eigen::MatrixXd pp(4, 4);
+            pp << rr(0, 0), rr(0, 1), rr(0, 2), pose[3], rr(1, 0), rr(1, 1), rr(1, 2), pose[4], rr(2, 0), rr(2, 1), rr(2, 2), pose[5], 0, 0, 0, 1;
+            Eigen::Vector4d p = {init_trans[3], init_trans[4], init_trans[5], 1.0};
+            p = init.inverse() * pp * p;
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp]: Reuse the same code twice (see descriptors.hpp, line 80) -> create a function/method ?

**Comment by dogoepp** [comment]
The variable names need clarification.

It could be also useful to clarify that the first part of the Vector6d are angle-axis representation of the rotations, and the second part are the positions.

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp [hunk 6]
```diff
@@ -161,16 +159,26 @@ namespace hexapod_dart {
...
+            _final_pos = Eigen::Vector3d(p(0), p(1), p(2));
 
             _covered_distance = std::round(_final_pos(0) * 100) / 100.0;
 
+            // Angle computation
+            _final_rot = dart::math::matrixToEulerXYZ(ro.inverse() * rr);
+
             // roll-pitch-yaw
-            _arrival_angle = std::round(dart::math::matrixToEulerXYZ(dart::math::expMapRot(_final_rot))(2) * 100) / 100.0;
+            _arrival_angle = std::round(_final_rot(2) * 100) / 100.0;
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp:181]: What angle do you want to get here ?

File: hexapod_dart/include/hexapod_dart/descriptors.hpp (hunk 1 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/descriptors.hpp]: Can we avoid it? I mean from math perspective...

File: hexapod_dart/include/hexapod_dart/descriptors.hpp (hunk 3 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/descriptors.hpp:107]: Why does `init_trans` reads _init translation_ and not _init transformation_? I do mean the second, though..

File: hexapod_dart/include/hexapod_dart/hexapod.hpp (hunk 4 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/hexapod.hpp:78]: DART's `getPositions()` returns the positions of all of the generalized coordinates (thus including joint positions etc). For the pose, we only want `orientation` and `position` of the COM (indices 0-5).

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp (hunk 5 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp]: We could. Let me think about it..

**Comment by costashatz** [comment]
> The variable names need clarification.
> 
> It could be also useful to clarify that the first part of the Vector6d are angle-axis representation of the rotations, and the second part are the positions.

Will do..

File: hexapod_dart/include/hexapod_dart/descriptors.hpp (hunk 3 above)
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/descriptors.hpp:107]: You're right, I should think the way you do :)

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp (hunk 6 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp:181]: Basically, it's just yaw angle of the orientation (relative to the initial orientation).

**Comment by costashatz** [comment]
@dogoepp check if it is ok...

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp [hunk 7]
```diff
@@ -198,32 +198,32 @@ namespace hexapod_dart {
...
         }
 
-        Eigen::Vector3d final_pos()
+        Eigen::Vector3d final_pos() const
         {
             return _final_pos;
         }
 
-        Eigen::Vector3d final_rot()
+        Eigen::Vector3d final_rot() const
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp]: Do you think it should be worth returning a const reference for the Vector3d methods ? Depends on the cost of the copy constructor.

File: hexapod_dart/include/hexapod_dart/hexapod.hpp [hunk 8]
```diff
@@ -117,8 +120,7 @@ namespace hexapod_dart {
 
             for (size_t i = 0; i < to_remove.size(); i++) {
                 auto tmp = _skeleton->getJoint(to_remove[i] - i * 3)->getChildBodyNode();
-                tmp->removeAllCollisionShapes();
-                tmp->removeAllVisualizationShapes();
+                tmp->removeAllShapeNodes();
```
Comment (dogoepp) [inline hexapod_dart/include/hexapod_dart/hexapod.hpp:123]: Is this part of the transformation fix ? What's happening here ?

File: hexapod_dart/src/test.cpp [hunk 9]
```diff
@@ -5,7 +5,7 @@ int main()
 {
     std::vector<int> brk = {};
 
-    auto global_robot = std::make_shared<hexapod_dart::Hexapod>("/home/kchatzil/Workspaces/ResiBots/source/sferes2/pexod.urdf", brk);
+    auto global_robot = std::make_shared<hexapod_dart::Hexapod>("/home/kchatzil/Workspaces/ResiBots/share/hexapod_models/URDF/pexod.urdf", brk);
```
Comment (dogoepp) [inline hexapod_dart/src/test.cpp]: This isn't very portable. Maybe we don't care ?

Comment (costashatz) [inline hexapod_dart/src/test.cpp]: Yeap.. We don't care...

File: hexapod_dart/include/hexapod_dart/hexapod.hpp (hunk 8 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/hexapod.hpp:123]: No.. These are changes for the new dart master...

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp (hunk 7 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp]: Yeap..we should..although we do not care that much.. I will do it...

File: hexapod_dart/src/test.cpp (hunk 9 above)
Comment (costashatz) [inline hexapod_dart/src/test.cpp]: > This isn't very portable. Maybe we don't care ?

Fixed this.. More portable now..

File: hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp (hunk 7 above)
Comment (costashatz) [inline hexapod_dart/include/hexapod_dart/hexapod_dart_simu.hpp]: > Do you think it should be worth returning a const reference for the Vector3d methods ? Depends on the cost of the copy constructor.

Did this too..

**Comment by dogoepp** [comment]
Assuming it all works, **OK** for me.

**Comment by costashatz** [comment]
> Assuming it all works, OK for me.

It's all about assumptions at the end the day.. :smile:

**Comment by dogoepp** [comment]
:)

**Comment by costashatz** [comment]
**Merging...**

**Pull request merged**